### ISSUE TITLE
feat: justrank service crate (POST /v1/rerank, sibling to justembed)

### ISSUE THREAD

**Issue by sweengineeringlabs** [sweengineeringlabs/rust-deeplearning#5]
## Summary

Create `justrank/` — a service crate sibling to `justembed/` that exposes
`POST /v1/rerank` for cross-encoder + (later) LLM-native reranking.
Mirrors justembed's two-crate workspace pattern (`conversion/` library
+ `systemd/` axum service), reuses the entire `swe-llmmodel-*` stack,
and follows `swearchitect/main/policy/architecture.policy.toml`.

## Relationship to other issues

- **Blocker for full v1 acceptance**: #4 (architecture readiness for
  bge-reranker-v2-m3 — `arch-bert`/`arch-xlm-roberta`, classifier-head
  support on `LanguageModel`, GGUF coverage, `PoolingStrategy::Cls`).
  This service issue can land scaffolding + a stub backend ahead of
  #4 closing; the real model load happens once #4 is complete.
- Downstream consumers (separate issues, not blocked on this issue
  beyond having a callable endpoint): swerag `SweRerankClient`,
  llmboot `RagAgentConfig.rerank` toggle.

## Workspace layout (mirrors justembed)

```
machinelearning/justrank/
├── Cargo.toml                       # workspace, members = [scoring, systemd]
├── README.md
├── scoring/                         # analogue of justembed/conversion/
│   ├── Cargo.toml
│   └── src/
│       ├── lib.rs                   # mod api; mod core; mod saf; pub use saf::*;
│       ├── api/
│       │   ├── mod.rs
│       │   ├── error.rs             # ScoringError, ScoringResult
│       │   ├── traits.rs            # pub trait Rank, pub trait ClassifierHead
│       │   └── types.rs             # RerankRequest, RerankResponse, ScoredDocument
│       ├── core/
│       │   ├── mod.rs
│       │   ├── classifier_head.rs   # impl ClassifierHead for DefaultClassifierHead
│       │   └── pair_format.rs       # impl pair tokenization helpers (CLS q SEP d SEP)
│       └── saf/
│           └── mod.rs               # pub use api::*; factory functions
└── systemd/                         # analogue of justembed/systemd/
    ├── Cargo.toml
    └── src/
        ├── lib.rs
        ├── api/
        │   ├── mod.rs
        │   ├── config.rs            # RerankServiceConfig (TOML-driven)
        │   └── error.rs             # RerankApiError
        ├── core/
        │   ├── mod.rs
        │   ├── config.rs            # impl from_config (TOML loader)
        │   ├── loader.rs            # GGUF -> RerankState
        │   ├── router.rs            # build_rerank_router, /v1/rerank, /health
        │   └── state.rs             # RerankState { model, tokenizer, model_id }
        ├── bin/
        │   └── rank.rs              # binary entry — argparse, tracing, serve
        └── saf/
            └── mod.rs               # pub use api::*; factory functions
```

Why the SEA layout matches the policy: `api/` holds declarations only
(rule 160), `core/` holds impl blocks only (rule 86, no traits in
core), `saf/` is the facade with no `pub use crate::core::*`
(rule 47), `spi/` is intentionally absent because there are no
external plug-in trait declarations to host — same pattern as
justembed.

## API contract

`POST /v1/rerank`

Request:
```json
{
  "query": "what causes intermittent connection drops?",
  "documents": [
    "Connection drops are typically caused by ...",
    "The OSI model defines seven layers ...",
    "TCP retransmission timers handle ..."
  ],
  "top_n": 2,
  "return_documents": false
}
```

Response:
```json
{
  "object": "list",
  "model": "bge-reranker-v2-m3-q4_k_m",
  "data": [
    {"index": 0, "relevance_score": 0.94},
    {"index": 2, "relevance_score": 0.71}
  ],
  "usage": {"total_tokens": 312}
}
```

`GET /health` returns `{"status": "ok", "model": "<id>"}` matching
justembed's shape.

## Backend strategy

**Two backends behind one router**, dispatched at startup based on
loaded model architecture:

1. **Encoder cross-encoder backend** (v1 — ships when #4 lands).
   Tokenizes `[CLS] query [SEP] doc [SEP]` per pair, runs encoder
   forward via `model.encode(...)` (or whatever the new
   sequence-classifier API is named), reads CLS hidden state,
   projects through the classifier head, returns scalar scores.

2. **Decoder LLM-native backend** (v2 — ships when #4 v2 architectures
   land). Builds a "is this relevant? answer yes/no:" prompt per pair,
   calls `LanguageModel::forward()`, extracts last-position logits,
   softmax over the {yes, no} subset, returns the yes-probability as
   the score.

**Initial PR (scaffold): stub backend** that accepts requests and
returns deterministic placeholder scores so swerag can integrate
end-to-end before real models load. The stub is gated behind a
config flag; loader.rs panics on real model load until #4 lands.

### Scaffold (this issue)

- Workspace builds clean against current `swe-llmmodel-*` deps
- `POST /v1/rerank` accepts the documented request shape and returns
  the documented response shape (stub returns placeholder scores)
- `GET /health` reports model id from config
- Config-driven (TOML), matching the project pattern. No `Default`
  impl with literal numbers; service ships a `config/default.toml`.
- All 5 SEA-layer rules (47/50/53/86/160) clean — verified manually
  against `swearchitect/main/policy/architecture.policy.toml`
- README documents the stub status and the path to v1

### v1 (depends on #4)

- Loader resolves a bge-reranker-v2-m3 GGUF and constructs the
  encoder + classifier-head model
- `/v1/rerank` returns scores from the real model
- Parity test: 50 (query, doc) pairs vs reference within tolerance
- Latency baseline reported (50 docs, CPU)

### v2 (later)

- Decoder backend wired behind the same `/v1/rerank` endpoint
- Backend selection automatic from loaded model architecture

## Out of scope

- Architecture readiness — covered by #4
- swerag integration — separate issue in the rag repo
- llmboot YAML config — separate issue in llmboot, blocked on swerag

## References

- Architecture template: `justembed/{conversion,systemd}/`
- Architecture policy: `swearchitect/main/policy/architecture.policy.toml`
- Architecture readiness: #4

**Comment by sweengineeringlabs** [comment]
Closing — wrong repo. Reranker as a service belongs alongside `swe-bm25` / `swe-rrf` / `swerag` / `swevecdb` (each its own repo under sweengineeringlabs), not as a sibling of `justembed` here. Will refile against the correct repo. Issue #4 (architecture readiness for cross-encoder + LLM-native rerankers in `swe-llmmodel-arch-*`) stays here — that work legitimately lives in the machinelearning repo.
